File: _add-two-numbers.js
/**
 * Definition for singly-linked list.
 * function ListNode(val, next) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.next = (next===undefined ? null : next)
 * }
 */
/**
 * @param {ListNode} l1
 * @param {ListNode} l2
 * @return {ListNode}
 */
var addTwoNumbers = function(l1, l2) {

    let res = new ListNode(0)
    let head = res
    let carry = 0

    while(l1 || l2){
        let val1=0
        let val2=0

        if(l1)
            val1 = l1.val
        
        if(l2)
            val2 = l2.val
        
        let sum = val1 + val2 + carry

        carry = Math.floor(sum/10)
        head.next = new ListNode(sum%10)

        if(l1)
            l1 = l1.next
        if(l2)
            l2 = l2.next
        
        head= head.next
    }

    if(carry>0){
        head.next = new ListNode(carry)
    }

    return res.next
    
};

File: Python/add-two-numbers.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    def addTwoNumbers(self, l1: Optional[ListNode], l2: Optional[ListNode]) -> Optional[ListNode]:

        head = ListNode()
        cur = head
        
        carry = 0

        while(l1!=None and l2!=None):
            val1 = l1.val
            val2 = l2.val
            newVal = val1 + val2
            if(carry == 1):
                newVal = newVal +1
                carry = 0
            
            if(newVal>9):
                carry =1
                newVal = newVal - 10
            
            newNode = ListNode(newVal)
            # print('newNode',newNode)

            cur.next = newNode
            cur = cur.next
            l1 = l1.next
            l2 = l2.next
        
        while(l1!=None):
            # print('l1 call',l1)
            val1 = l1.val

            if(carry == 1):
                val1 = val1 +1
                carry = 0
            
            if(val1>9):
                carry =1
                val1 = val1 - 10
            
            newNode=ListNode(val1)
            # print('newNode',newNode)
            cur.next = newNode
            cur = cur.next
            l1 = l1.next
        
        while(l2!=None):
            # print('l2 call',l2)

            val1 = l2.val

            if(carry == 1):
                val1 = val1 +1
                carry = 0
            
            if(val1>9):
                carry =1
                val1 = val1 - 10
            
            newNode=ListNode(val1)
            # print('newNode',newNode)
            cur.next = newNode
            cur = cur.next
            l2 = l2.next
        
        if(carry ==1):
            # print('carry check')
            newNode=ListNode(1)
            cur.next = newNode

        return head.next


# 9,9,9,9,9,9,9
# 9,9,9,9

# 8 9 9 9 0 0 0 1
#   1 1 1 1 1 1 1















